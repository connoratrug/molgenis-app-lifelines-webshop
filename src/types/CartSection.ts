import { Section } from './Section'
import { Variable } from './Variable'

interface CartVariable extends Variable {
  subsection: number
}

interface CartSubsection {
  name: string
  variables: CartVariable[]
}

export default interface CartSection extends Section {
  subsections: CartSubsection[]
}
