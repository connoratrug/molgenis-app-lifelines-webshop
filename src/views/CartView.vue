<template>
  <div id="cart-view" class="row">
      <div class="col">
        <h3 class="mb-3">3. Order variables</h3>
        <template v-if="selectedVariableIds.length">
           <h5>Selected variables</h5>
           <spinner-animation v-show="loading"/>
           <ul>
             <li
              v-for="selectedVariable in selectedVariables"
              :key="selectedVariable.id">
              <template v-if="selectedVariable.label">{{selectedVariable.label}}</template>
              <template v-else>{{selectedVariable.name}}</template>
               <span> {{ varAssesments(selectedVariable.id) }}</span>
              </li>
           </ul>
        </template>
         <template v-else>
            <h5>No variables selected</h5>
            <p>Use the shop tab to select variables to order</p>
        </template>

      </div>
  </div>
</template>

<script>
import Vue from 'vue'
import SpinnerAnimation from '../components/animations/SpinnerAnimation.vue'
import variableRepository from '../repository/variablesRepository.ts'
import assessmentsRepository from '../repository/assessmentsRepository.ts'

export default Vue.extend({
  name: 'CartView',
  components: { SpinnerAnimation },
  data: function () {
    return {
      selectedVariables: [],
      assessments: [],
      loading: false
    }
  },
  computed: {
    gridSelection () {
      return this.$store.state.gridSelection
    },
    selectedVariableIds () {
      return Object.keys(this.gridSelection)
    },
    variableAssesments () {
      let assessments = {}

      for (const [variableId, assessmentIds] of Object.entries(this.gridSelection)) {
        const assessmentNames = assessmentIds.map(assessmentId => this.assessments[assessmentId].name)
        assessments[variableId] = '( ' + assessmentNames.join(',') + ' )'
      }

      return assessments
    }
  },
  methods: {
    varAssesments (variableId) {
      return this.variableAssesments[variableId]
    }
  },
  mounted () {
    this.loading = true
    variableRepository.resolveVariableIds(this.selectedVariableIds).then((vars) => {
      this.loading = false
      this.selectedVariables = vars
    })

    assessmentsRepository.getAssessments().then((assessments) => {
      this.assessments = assessments
    })
  }
})
</script>
