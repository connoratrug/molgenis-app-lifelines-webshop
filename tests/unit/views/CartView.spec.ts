import { mount, createLocalVue, RouterLinkStub } from '@vue/test-utils'

import CartSection from '@/types/CartSection'
import CartView from '@/views/CartView.vue'
import Vuex from 'vuex'
import Vue from 'vue'
import flushPromises from 'flush-promises'
import BootstrapVue from 'bootstrap-vue'

Vue.filter('i18n', (value: string) => value) // Add dummy filter for i18n

describe('CartView.vue', () => {
  const localVue = createLocalVue()
  localVue.use(Vuex)
  localVue.use(BootstrapVue)

  let stubs = {
    RouterLink: RouterLinkStub
  }

  let mocks = {
    $router: {
      push: jest.fn()
    }
  }

  let store: any
  let actions: any

  let cartTree: CartSection[]

  const getters = {
    cartTree: () => cartTree,
    isSignedIn: () => true
  }

  beforeEach(() => {
    let state: any
    actions = {
      save: jest.fn()
    }

    state = {
      gridSelection: {
        123: [1, 2],
        456: [3]
      },
      variables: {
        123: {
          id: 123,
          name: 'var123',
          label: 'var 123'
        },
        456: {
          id: 456,
          name: 'var456',
          label: 'var 456'
        }
      },
      assessments: {
        1: {
          id: 1,
          name: 'assessment1'
        },
        2: {
          id: 1,
          name: 'assessment3'
        },
        3: {
          id: 1,
          name: 'assessment3'
        }
      }
    }

    cartTree = [{
      id: 1,
      name: 'Section 1',
      subsections: [{
        name: 'Subsection 1',
        variables: [{
          id: 123,
          name: 'var123',
          label: 'var 123',
          subsections: [1],
          subsection: 1
        }]
      }]
    }, {
      id: 2,
      name: 'Section 2',
      subsections: [{
        name: 'Subsection 2',
        variables: [{
          id: 456,
          name: 'var456',
          label: 'var 456',
          subsections: [2],
          subsection: 2
        }]
      }]
    }]

    store = new Vuex.Store({
      state,
      actions,
      getters
    })
  })

  it('renders cart view', () => {
    const wrapper = mount(CartView, { stubs, store, localVue })
    expect(wrapper.findAll('li').at(0).text()).toEqual('var 123 ( assessment1, assessment3 )')
    expect(wrapper.findAll('li').at(1).text()).toEqual('var 456 ( assessment3 )')
    expect(wrapper.find('#cart-view')).toMatchSnapshot()
  })

  it('renders a save button that saves the current state', async () => {
    actions.save.mockResolvedValue('12345')
    const wrapper = mount(CartView, { stubs, store, localVue, mocks })
    wrapper.find('.save').trigger('click')
    await flushPromises()
    expect(mocks.$router.push).toHaveBeenCalledWith({ name: 'load', params: { orderNumber: '12345' } })
  })
})
