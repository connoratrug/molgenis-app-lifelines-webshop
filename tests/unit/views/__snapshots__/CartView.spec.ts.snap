// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartView.vue renders cart view 1`] = `
<div id="cart-view">
  <h3 class="h4">lifelines-webshop-cart-header</h3>
  <!---->
  <div role="tablist" class="mt-3 mb-3"><button type="button" class="btn btn-primary save">lifelines-webshop-save-btn-label</button> <button type="button" class="btn btn-warning ml-2">
      lifelines-webshop-order-btn-label
    </button>
    <h4 class="h5 mt-3">lifelines-webshop-cart-selected-list-header</h4>
    <div class="card mb-1">
      <!---->
      <!---->
      <header role="tab" class="card-header p-1"><a target="_self" href="#" class="btn btn-info btn-block" aria-controls="accordion-0" aria-expanded="false" role="button">Section 1</a></header>
      <div id="accordion-0" class="collapse show" role="tabpanel">
        <div class="card-body">
          <!---->
          <!---->
          <div>
            <h5 class="h6">Subsection 1</h5>
            <ul>
              <li><span>var 123 ( assessment1, assessment3 )</span></li>
            </ul>
          </div>
        </div>
      </div>
      <!---->
      <!---->
    </div>
    <div class="card mb-1">
      <!---->
      <!---->
      <header role="tab" class="card-header p-1"><a target="_self" href="#" class="btn btn-info btn-block" aria-controls="accordion-1" aria-expanded="false" role="button">Section 2</a></header>
      <div id="accordion-1" class="collapse" style="display: none;" role="tabpanel">
        <div class="card-body">
          <!---->
          <!---->
          <div>
            <h5 class="h6">Subsection 2</h5>
            <ul>
              <li><span>var 456 ( assessment3 )</span></li>
            </ul>
          </div>
        </div>
      </div>
      <!---->
      <!---->
    </div>
  </div>
  <!---->
</div>
`;
